docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+
+An orphan object is any object which is no longer referenced by a running node or by metadata.
+
+Examples of how orphan objects arise:
+
+- A node is doing compaction and writes a layer object, then crashes before it writes the
+  index_part.json that references that layer.
+- A partition node carries on running for some time, and writes out an unbounded number of
+  objects while it believes itself to be the rightful writer for a tenant.
+
+Orphan objects are functionally harmless, but have a small cost due to S3 capacity consumed. We
+may clean them up at some time in the future, but doing a ListObjectsv2 operation and cross
+referencing with the latest metadata to identify objects which are not referenced. This kind
+of deletion would go through the same deletion queue as other object deletions, and be subject
+to the same checks for generation number freshness.
+
+### Optional: Safekeeper optimization for stale pageservers
+
+As an optimization, we may extend the safekeeper to be asynchronously updated about pageserver node
+generation numbers. We may do this by including the generation in messages from pageserver to safekeeper,
+and having the safekeeper write the highest generation number it has seen for each node to the storage broker.
+
+Once the safekeeper has visibility of the most recent generation, it may reject requests from pageservers
+with stale generation numbers: this would reduce any possible extra load on the safekeeper from stale pageservers,
+and provide feedback to stale pageservers that they should shut down.
+
+This is not required for safety: reads from a stale pageserver are functionally harmless and only
+waste system resources. Logical writes (i.e. updates to remote_persistent_lsn that can cause trimming)
+are handled on the pageserver side (see "WAL trim changes" above).
+
+Note that the safekeeper should only reject reads for stale _pageserver_ generations. Stale _attachment_
+generations are valid for reads, as a tenant may be multi-attached during a migration, where two different
+pageservers are both replaying the WALs for the same tenant.
+
+## Operational impact
+
+### Availability
+
+Coordination of generation numbers via the control plane introduce a dependency for certain
+operations:
+
+1. Starting new pageservers (or activating pageservers after a restart)
+2. Executing enqueued deletions
+3. Advertising updated `remote_consistent_lsn` to enable WAL trimming
+
+Item 1. would mean that some in-place restarts that previously would have resumed service even if the control plane were
+unavailable, will now not resume service to users until the control plane is available. We could
+avoid this by having a timeout on communication with the control plane, and after some timeout,
+resume service with the node's previous generation (assuming this was persisted to disk). However,
+this is unlikely to be needed as the control plane is already an essential & highly available component.
```

Comment:
The idea of a time out & resume with previous generation tripped me up. 
I thought earlier we said that the node generation ID uniquely identifies the process (as opposed to node ID identifying essentially the disk state).

Reply:
If we implemented this, it would indeed violate that invariant that 1 generation == 1 process.  It would be safe, but ugly.  I've added a line to clarify that & that I don't want to do it :-)  3e47d681